src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -233,29 +240,46 @@
                        (update-report uuid stamp))
         factset (some-> factset
                         (update-factset rand-percentage stamp))]
-    (when catalog (>!! command-send-ch [:catalog host 9 catalog]))
-    (when report (>!! command-send-ch [:report host 8 report]))
-    (when factset (>!! command-send-ch [:factset host 5 factset]))
+
+    (when catalog
+      (when-not (>!! command-send-ch [:catalog host 9 catalog])
+        (throw+ channel-closed-map)))
+
+    (when report
+      (when-not (>!! command-send-ch [:report host 8 report])
+        (throw+ channel-closed-map)))
+
+    (when factset
+      (when-not (>!! command-send-ch [:factset host 5 factset])
+        (throw+ channel-closed-map)))
 
     (assoc state
            :catalog catalog
            :factset factset)))
 
-(defn submit-n-messages
-  "Given a list of host maps, send `num-messages` to each host.  The function
-   is recursive to accumulate possible catalog mutations (i.e. changing a previously
-   mutated catalog as opposed to different mutations of the same catalog)."
+(defn create-messages
+  "Given a list of host maps, enqueue `num-messages` into
+  `command-send-ch` for each host.  The function is recursive to
+  accumulate possible catalog mutations (i.e. changing a previously
+  mutated catalog as opposed to different mutations of the same
+  catalog)."
   [hosts num-msgs rand-percentage command-send-ch]
-  (log/info
-   (trs "Sending {0} messages for {1} hosts, will exit upon completion"
-        num-msgs (count hosts)))
-  (loop [mutated-hosts hosts
-         msgs-to-send num-msgs
-         stamp (time/minus (time/now) (time/minutes (* 30 num-msgs)))]
-    (when-not (zero? msgs-to-send)
-      (recur (mapv #(update-host % rand-percentage stamp command-send-ch) mutated-hosts)
-             (dec msgs-to-send)
-             (time/plus stamp (time/minutes 30))))))
+  {:run-task (fn []
+               (println-err
+                (trs "Sending {0} messages for {1} hosts, will exit upon completion"
+                     num-msgs (count hosts)))
+               (loop [mutated-hosts hosts
+                      msgs-to-send num-msgs
+                      stamp (time/minus (time/now) (time/minutes (* 30 num-msgs)))]
+                 (if (zero? msgs-to-send)
+                   (close! command-send-ch)
+                   (let [mutated-hosts (mapv #(update-host % rand-percentage stamp command-send-ch) mutated-hosts)]
+                     (recur mutated-hosts
+                            (dec msgs-to-send)
+                            (time/plus stamp (time/minutes 30)))))))
+   :stop-task (fn [] (close! command-send-ch))
```

Comment:
hm. The normal way to stop a loop that's putting stuff onto a channel is to close the channel. Then the loop should check if >! (>!!) returns false and break out.